### PR TITLE
Unblock the hooks execution

### LINKED ISSUE #11231: Reduce time spent when adding runtime hooks
As visible in the following profiling screenshot:
Screenshot from 2024-10-02 12-27-23
we spend 2 seconds just waiting to add a visualization update hook. 

This is because runtime hooks synchronize on the list of hooks while executing them, thus blocking any further updates. We could be a bit smarter when it comes to locking in this case e.g., fetching list of hooks in a synchronized block, executing them outside of the block and checking if new hooks were added at the end?

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #11231 

Changelog:
- fix: do not block during the hooks execution

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java [hunk 1]
```diff
@@ -21,15 +22,26 @@ public void add(Runnable hook) {
 
   @Override
   public void run() {
+    Set<Runnable> hooksToRun;
     synchronized (hooks) {
-      for (Runnable hook : hooks) {
+      hooksToRun = new LinkedHashSet<>(hooks);
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java]: I rather use `hooks.toArray(new Runnable[0])` and then immediately `hooks.clear()`.

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java [hunk 2]
```diff
@@ -21,15 +22,26 @@ public void add(Runnable hook) {
...
         } catch (Exception e) {
           logger.error("Failed to run execution hook.", e);
+        } finally {
+          hooksExecuted.add(hook);
         }
       }
-      hooks.clear();
+    } finally {
+      synchronized (hooks) {
+        hooks.removeAll(hooksExecuted);
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java]: Unless there is somebody who checks `hooks.isEmpty()` or `hooks.size()` and basis some decision upon such check, then I suggest to `hooks.clear()` on line 28 and remove this `sychronized` block altogether.

**Review by JaroslavTulach — APPROVED** [review]
The change is logically fine. I suggest to call `clear()` in the first `synchronized` block and remove the second one completely.

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java (hunk 2 above)
Comment (4e6) [reply engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java]: I realized that I can simplify the logic :+1:

**Pull request merged**